instrumentation/vertx/vertx-kafka-client-3.6/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/vertx/kafka/v3_6/InstrumentedBatchRecordsHandler.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.javaagent.instrumentation.vertx.kafka.v3_6;
+
+import static io.opentelemetry.javaagent.instrumentation.vertx.kafka.v3_6.VertxKafkaSingletons.batchProcessInstrumenter;
+
+import io.opentelemetry.context.Context;
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.instrumentation.api.field.VirtualField;
+import io.opentelemetry.javaagent.bootstrap.kafka.KafkaClientsConsumerProcessTracing;
+import io.vertx.core.Handler;
+import javax.annotation.Nullable;
+import org.apache.kafka.clients.consumer.ConsumerRecords;
+
+public final class InstrumentedBatchRecordsHandler<K, V> implements Handler<ConsumerRecords<K, V>> {
+
+  private final VirtualField<ConsumerRecords<K, V>, Context> receiveContextField;
+  @Nullable private final Handler<ConsumerRecords<K, V>> delegate;
+
+  public InstrumentedBatchRecordsHandler(
+      VirtualField<ConsumerRecords<K, V>, Context> receiveContextField,
+      @Nullable Handler<ConsumerRecords<K, V>> delegate) {
+    this.receiveContextField = receiveContextField;
+    this.delegate = delegate;
+  }
+
+  @Override
+  public void handle(ConsumerRecords<K, V> records) {
+    Context parentContext = getParentContext(records);
+
+    if (!batchProcessInstrumenter().shouldStart(parentContext, records)) {
+      callDelegateHandler(records);
+      return;
+    }
+
+    // the instrumenter iterates over records when adding links, we need to suppress that
+    boolean previousWrappingEnabled = KafkaClientsConsumerProcessTracing.setEnabled(false);
+
+    Context context = batchProcessInstrumenter().start(parentContext, records);
+    Throwable error = null;
+    try (Scope ignored = context.makeCurrent()) {
+      callDelegateHandler(records);
+    } catch (Throwable t) {
+      error = t;
+      throw t;
+    } finally {
+      try {
+        batchProcessInstrumenter().end(context, records, null, error);
+      } finally {
+        KafkaClientsConsumerProcessTracing.setEnabled(previousWrappingEnabled);
+      }
+    }
```

Comment:
this is the pattern that makes me feel safest for thread locals

```suggestion
    boolean previousWrappingEnabled = KafkaClientsConsumerProcessTracing.setEnabled(false);
    try {
      Context context = batchProcessInstrumenter().start(parentContext, records);
      Throwable error = null;
      try (Scope ignored = context.makeCurrent()) {
        callDelegateHandler(records);
      } catch (Throwable t) {
        error = t;
        throw t;
      } finally {
        batchProcessInstrumenter().end(context, records, null, error);
      }
    } finally {
      KafkaClientsConsumerProcessTracing.setEnabled(previousWrappingEnabled);
    }
```